library/bigip_software.py

```diff
@@ -212,673 +290,1060 @@
         content = io.BytesIO(content)
-
         context = etree.iterparse(content)
         for action, elem in context:
             if elem.text:
                 text = elem.text
-
-            if elem.tag == 'productName':
-                result['product'] = text
-            elif elem.tag == 'version':
-                result['version'] = text
+            if elem.tag == 'version':
+                self._values['version'] = text
             elif elem.tag == 'buildNumber':
-                result['build'] = text
+                self._values['build'] = text
 
-        return result
-
-    def flush(self):
-        result = dict()
-        state = self.params['state']
-        volume = self.params['volume']
-        software = self.params['software']
-        reuse_inactive_volume = self.params['reuse_inactive_volume']
-
-        if state == 'activated' or state == 'installed':
-            if not volume:
-                if not reuse_inactive_volume:
-                    raise F5ModuleError("You must specify a volume")
-            elif not software:
-                raise F5ModuleError("You must specify a base image")
-
-        if state == "activated":
-            changed = self.activated()
-        elif state == "installed":
-            changed = self.installed()
-        elif state == "present":
-            changed = self.present()
-        elif state == "absent":
-            changed = self.absent()
-
-        if state in ['activated', 'installed', 'present']:
-            if not self.params['check_mode']:
-                current = self.read()
-                result.update(current)
-
-        changed = True
-        result.update(dict(changed=changed))
-        return result
+    def _find_iso_content(self, iso):
+        paths = ['/METADATA.XML', 'metadata.xml']
 
+        for path in paths:
+            try:
+                content = iso.record(path).content
+                return content
+            except KeyError:
+                pass
+        raise Exception(
+            "Unable to find metadata file in ISO. Please file a bug."
+        )
 
-class BigIpSoapApi(BigIpCommon):
-    def __init__(self, *args, **kwargs):
-        super(BigIpSoapApi, self).__init__(*args, **kwargs)
+    def handle_remote_source(self, link):
+        source = urlparse.urlparse(link)
+        name = source.path.split('/')[-1]
+        if source.scheme:
+            if source.scheme == 'http':
+                return name
+            elif source.scheme == 'https':
+                self._values['secure'] = True
+                return name
+            else:
+                err = 'Only remote HTTP or HTTPS sources are supported.'
+                raise F5ModuleError(err)
 
-        self.api = bigip_api(kwargs['server'],
-                             kwargs['user'],
-                             kwargs['password'],
-                             kwargs['validate_certs'])
+    def _list_volumes(self):
```

Comment:
since these are methods whose sole purpose is to communicate with the device, the pattern we use for that is naming them *_on_device or *_from_device. I would change these to match that pattern